Actor State TTL

JoshVanL opened the pull request:
Dapr Actor state TTL is a feature that was added in Dapr 1.12.

Adds support for Actor State TTL. TTL of actor state is set by using the `ttlInSeconds` metadata option when saving actor state. The actor state cache uses the `ttlExpireTime` return metadata field to determine whether the key is still valid, which is important as the cache may become populated with `TTL` keys which were not created by the current manager.

SDK developers are encouraged to use TTLs when storing any actor state in order to prevent stale data accumulating in the actor state store over time.

Closes https://github.com/dapr/dotnet-sdk/pull/1164

JoshVanL commented on src/Dapr.Actors/DaprHttpInteractor.cs:
```diff
@@ -57,7 +57,7 @@ internal class DaprHttpInteractor : IDaprInteractor
         }
 
-        public async Task<string> GetStateAsync(string actorType, string actorId, string keyName, CancellationToken cancellationToken = default)
+        public async Task<ActorStateResponse<string>> GetStateAsync(string actorType, string actorId, string keyName, CancellationToken cancellationToken = default)
```
We want to keep `ActorStateResponse` as an internal type. How do we keep complication happy with this constraint?

JoshVanL replied:
@philliphoff any ideas here?

philliphoff replied:
`ActorStateResponse` is already `internal`, no, or are you referring to the fact that this method is `public`? `IDaprHttpInteractor` and `DaprrHttpInteractor` are already `internal` types, so nothing of them can be seen outside the SDK.  The `public` on the method only applies to other types within the SDK.  (Method visibility is always masked by visibility of the underlying type.)

philliphoff commented on src/Dapr.Actors/Runtime/ActorStateManager.cs:
```diff
@@ -35,17 +36,17 @@ internal ActorStateManager(Actor actor)
         }
 
-        public async Task AddStateAsync<T>(string stateName, T value, CancellationToken cancellationToken)
+        public async Task AddStateAsync<T>(string stateName, T value, CancellationToken cancellationToken, int? ttlInSeconds)
```
Would it make sense to instead add an overload for when the TTL is specified:
`AddStateAsync<T>(string stateName, T value, int ttlInSeconds)` rather than deviate from the .NET guidance that `CancellationToken` be the *last* parameter of a method declaration.  An exception exists to maintain backward compatibility with existing public APIs, but an overload would seem to provide the same effect while maintaining a consistent API.

Also, is it important for TTL to be specified in integer seconds as opposed to, say, a `TimeSpan`, which might be more usable for larger values?  (What do we expect the typical TTL to be?)

philliphoff replied:
Right. I was think more of the actual values used for TTLs.  For example, the samples specify a TTL of `360`.  Would that be better expressed as `TimeSpan.FromMinutes(3)` so that users need not do the math (and potentially get it wrong), specially if they have much larger TTLs (e.g. hours/days).

JoshVanL replied:
Gotcha, yes that makes sense I'll add that in.

philliphoff commented on src/Dapr.Actors/Communication/ActorStateResponse.cs:
```diff
@@ -0,0 +1,50 @@
+        /// </summary>
+        /// <param name="value">The response value.</param>
+        /// <param name="ttlExpireTime">The time to live expiration time.</param>
+        public ActorStateResponse(T value, DateTime? ttlExpireTime)
```
Consider using `DateTimeOffset` to ensure comparability of expiration times?

JoshVanL replied:
Done :+1:

philliphoff commented on src/Dapr.Actors/Runtime/ActorStateChange.cs:
```diff
@@ -68,5 +70,39 @@ public ActorStateChange(string stateName, Type type, object value, StateChangeKi
+        /// <remarks>
+        /// If null, the state will not expire.
+        /// </remarks>
+        public int? TTLInSeconds {
```
Since this conversion to seconds is only used in a single place (for the specific circumstances of the runtime API), would it make sense to move this logic there and only expose `TTLExpireTime`?  Is there value in exposing both to users of `ActorStateChange`? (Especially since this property is a constantly changing value, being relative to the current date/time, which users may not expect from a property.)

JoshVanL replied:
Thanks @philliphoff, I've updated the PR to only have hold the `TTLExpireTime` value which gets converted to int seconds when saving the actor state.

philliphoff commented on src/Dapr.Actors/Runtime/ActorStateManager.cs:
```diff
@@ -102,8 +144,8 @@ public async Task<ConditionalValue<T>> TryGetStateAsync<T>(string stateName, Can
-                // Check if the property was marked as remove in the cache
-                if (stateMetadata.ChangeKind == StateChangeKind.Remove)
+                // Check if the property was marked as remove in the cache or is expired
+                if (stateMetadata.ChangeKind == StateChangeKind.Remove || stateMetadata.TTLExpireTime <= DateTime.UtcNow)
```
Nit: `DateTimeOffset.UtcNow`?

I'm virtually certain that the built-in coersion between `DateTime` and `DateTimeOffset` make it equivalent, but perhaps best to use `DateTimeOffset` as consistently as possible.

philliphoff replied:
Also might be useful to keep a watch on .NET 8 which (finally) adds a [means of abstracting time](https://learn.microsoft.com/en-us/dotnet/api/system.timeprovider?view=net-8.0) to simplify testing.

JoshVanL replied:
Thanks! Done

philliphoff commented on src/Dapr.Actors/Runtime/ActorStateManager.cs:
```diff
@@ -405,9 +551,11 @@ private StateMetadata(object value, Type type, StateChangeKind changeKind)
-            public static StateMetadata Create<T>(T value, StateChangeKind changeKind)
+            public DateTimeOffset? TTLExpireTime { get; set; }
+
+            public static StateMetadata Create<T>(T value, StateChangeKind changeKind, DateTimeOffset? ttlExpireTime = null, TimeSpan? ttl = null)
```
Is the intent to have a method which allows a TTL, expiry time, or neither?  If so, consider three methods, one which doesn't accept either, one that accepts the offset, and a third that accepts the time span.  That would help ensure proper usage.

JoshVanL replied:
Yes, that makes sense- they are intended to be mutually exclusive. I've moved to using 3 functions.

philliphoff commented on src/Dapr.Actors/Runtime/ActorStateManager.cs:
```diff
@@ -392,11 +529,20 @@ private void EnsureStateProviderInitialized()
+                    throw new ArgumentException("Cannot specify both TTLExpireTime and TTL");
+                }
+                if (ttl.HasValue) {
+                    this.TTLExpireTime = DateTime.UtcNow.Add(ttl.Value);
```
`DateTimeOffset.UtcNow`?  (Doesn't really *need* to be specified in UTC, as sometimes keeping the local time is useful for logging).

philliphoff commented on src/Dapr.Actors/Runtime/DaprStateProvider.cs:
```diff
@@ -132,6 +133,12 @@ private async Task DoStateChangesTransactionallyAsync(string actorType, string a
                         }
+
+                        if (stateChange.TTLExpireTime.HasValue) {
+                            var ttl = (int)Math.Ceiling((stateChange.TTLExpireTime.Value - DateTime.UtcNow).TotalSeconds);
```
Just a question, but is it spec'd to use the ceiling (vs. the floor or rounding) for TTLs?

JoshVanL replied:
The spec is that it needs to be an integer value. It makes sense to me to be consistent in how we translate a fractional seconds. Rounding up a fractional second ensures that a value has a TTL of 1 second, rather than 0.

JoshVanL commented:
@philliphoff do you understand why the integration tests are failing? I'm not too sure what the problem is.

philliphoff commented:
> @philliphoff do you understand why the integration tests are failing? I'm not too sure what the problem is.

Dunno.  Maybe retry the job in case it was just a transient issue?

JoshVanL commented:
@philliphoff Doesn't seem to be the case. Struggling to understand what is going wrong here.. at the limit of my .NET knowledge :grimacing:

philliphoff commented on src/Dapr.Actors/DaprHttpInteractor.cs:
```diff
@@ -72,7 +72,18 @@ HttpRequestMessage RequestFunc()
             var stringResponse = await response.Content.ReadAsStringAsync();
-            return stringResponse;
+
+            var ttlExpireTime = new DateTime();
```
```suggestion
            DateTimeOffset? ttlExpireTime = null;
```

philliphoff commented on src/Dapr.Actors/Runtime/ActorStateManager.cs:
```diff
@@ -102,8 +144,8 @@ public async Task<ConditionalValue<T>> TryGetStateAsync<T>(string stateName, Can
-                // Check if the property was marked as remove in the cache
-                if (stateMetadata.ChangeKind == StateChangeKind.Remove)
+                // Check if the property was marked as remove in the cache or is expired
+                if (stateMetadata.ChangeKind == StateChangeKind.Remove || stateMetadata.TTLExpireTime <= DateTimeOffset.UtcNow)
```
```suggestion
                if (stateMetadata.ChangeKind == StateChangeKind.Remove || (stateMetadata.TTLExpireTime.HasValue && stateMetadata.TTLExpireTime <= DateTimeOffset.UtcNow))
```

philliphoff commented on src/Dapr.Actors/Runtime/ActorStateManager.cs:
```diff
@@ -114,10 +156,10 @@ public async Task<ConditionalValue<T>> TryGetStateAsync<T>(string stateName, Can
             if (conditionalResult.HasValue)
             {
-                stateChangeTracker.Add(stateName, StateMetadata.Create(conditionalResult.Value, StateChangeKind.None));
+                stateChangeTracker.Add(stateName, StateMetadata.Create(conditionalResult.Value.Value, StateChangeKind.None, ttlExpireTime: conditionalResult.Value.TTLExpireTime));
```
```suggestion
                stateChangeTracker.Add(stateName, StateMetadata.Create(conditionalResult.Value.Value, StateChangeKind.None, ttlExpireTime: conditionalResult.Value.TTLExpireTime));

                return new ConditionalValue<T>(true, conditionalResult.Value.Value);
```

philliphoff commented on src/Dapr.Actors/Runtime/DaprStateProvider.cs:
```diff
@@ -43,36 +44,36 @@ public DaprStateProvider(IDaprInteractor daprInteractor, JsonSerializerOptions j
+            var response = await this.daprInteractor.GetStateAsync(actorType, actorId, stateName, cancellationToken);
 
-            if (stringResult.Length != 0)
+            if (response.Value.Length != 0 && response.TTLExpireTime > DateTimeOffset.UtcNow)
```
```suggestion
            if (response.Value.Length != 0 && (!response.TTLExpireTime.HasValue || response.TTLExpireTime > DateTimeOffset.UtcNow))
```

philliphoff commented:
> @philliphoff Doesn't seem to be the case. Struggling to understand what is going wrong here.. at the limit of my .NET knowledge 😬

Poked around this morning and found a couple of issues where existing state was ignored in favor of default values (see above suggestions), which were causing the E2E to fail.  It might be good to add some unit tests for `DaprHttpInteractor`, `ActorStateManager`, and `DaprStateProvider`, if reasonably possible, to ensure that all the cases are covered (e.g. when values are cached vs. non-cached), pre- and post-expiration, etc.).

philliphoff reviewed (changes requested):
Needs some changes (added as suggestion comments) in order to resolve E2E failure.

JoshVanL commented:
Thank you @philliphoff, really appreciate your help with this!

I've gone ahead and added those fixes, expanded the E2E state tests and added unit tests for those classed. Please take another look :slightly_smiling_face:

JoshVanL commented:
Hi @philliphoff, please let me know what you think when you have time
